models/db/context.go

```diff
@@ -96,13 +102,32 @@ func TxContext() (*Context, Committer, error) {
 	return newContext(DefaultContext, sess, true), sess, nil
 }
 
+var ErrAlreadyInTransaction = errors.New("database connection has already been in a transaction")
+
 // WithTx represents executing database operations on a transaction
-// you can optionally change the context to a parent one
-func WithTx(f func(ctx context.Context) error, stdCtx ...context.Context) error {
-	parentCtx := DefaultContext
-	if len(stdCtx) != 0 && stdCtx[0] != nil {
-		// TODO: make sure parent context has no open session
-		parentCtx = stdCtx[0]
+func WithTx(parentCtx context.Context, f func(ctx context.Context) error) error {
+	if InTransaction(parentCtx) {
+		return ErrAlreadyInTransaction
+	}
+
+	sess := x.NewSession()
+	defer sess.Close()
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	if err := f(newContext(parentCtx, sess, true)); err != nil {
+		return err
+	}
+
+	return sess.Commit()
+}
+
+// MustTx represents executing database operations on a transaction, if the transaction exist,
+// this function will reuse it otherwise will create a new one and close it when finished.
+func MustTx(parentCtx context.Context, f func(ctx context.Context) error) error {
```

Comment:
I'm a bit confused about the purpose of `WithTx` and `MustTx`.
I prefer the other way around:
`WithTx` **always** executes the function in a database, no matter if the transaction exists or not.
`MustTx` requires that there is a transaction already, and
`MustNotTx` could also be present to require that there is no transaction yet.

Reply:
`WithTx` will always open a new transaction, if a transaction exist in context, return an error.
`MustTx` will try to open a new transaction if no transaction exist in context. That means it will always enter a transaction if there is no error.
I don't know where we need `MustNotTx`.